### DISCUSSION TITLE
Domain events and the "Unit of Work" pattern

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#71] [General]
# Domain events and the "Unit of Work" pattern

https://steven-giesel.com/blogPost/ae55581a-9722-4735-8d0e-bfcfe4f6ad5a

**Comment by fisnik97** [comment]
What happens when you involve some other service functionality inside the MediatR command handler after doing some changes on Db.
Let's say I've modified some data, and now I am publishing some integration events, when the events are published and the command handler completes the middleware will invoke the save changes... What if the Db fails to make the changes, or what if you need the data that you are saving for processing next something based on their results ?

**Reply by linkdotnet** [reply]
Hey @fisnik97,

the idea is that one request would mimic one transaction inside that one system. So if there are multiple systems involved, the approach described in the blog post will not work. Patterns like the outbox pattern would be the way to go here.

Nevertheless, let's assume we have one system.

Your scenario is that you get a request, where you change some data and one or several handlers also mutate their respective data.
Then basically on `SaveChangesAsync` inside your middleware, you get an exception (for example, due to some constraint violations). Did I recap that correctly?

If so - then all of your transactions are not committed and therefore, the whole state is like before the request itself.

> you are saving for processing next something based on their results 

If you use Entity Framework Core you have an implicit transaction open until you are calling `SaveChanges`. This is scoped to your `DbContext` - so as lons as you use the same `DbContext` throughout your request (which is the default lifetime scope if you use methods like `AddDbContext`) you see the uncommitted work.

If you don't use Entity Framework, you could manually open the transaction inside the Middleware and close the transaction after all your handlers are done.
```csharp
try
{
    OpenTransaction();
    await next(context);
    CommitTransaction();
}
catch (Exception)
{
    DisposeTransaction();
    _logger.LogError("An error occurred while saving the changes to the database.");
    throw;
}
```
